src/types/dict.rs

```diff
@@ -401,85 +428,303 @@ fn dict_len(dict: &Bound<'_, PyDict>) -> Py_ssize_t {
+                move |(), x| match f(x) {
+                    Some(x) => ControlFlow::Break(x),
+                    None => ControlFlow::Continue(()),
+                }
+            }
+
+            match iter.try_fold((), check(f)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
+
+    #[inline]
+    #[cfg(all(Py_GIL_DISABLED, not(feature = "nightly")))]
+    fn position<P>(&mut self, predicate: P) -> Option<usize>
+    where
+        Self: Sized,
+        P: FnMut(Self::Item) -> bool,
+    {
+        self.with_critical_section(|iter| {
+            #[inline]
+            fn check<'a, T>(
+                mut predicate: impl FnMut(T) -> bool + 'a,
+                acc: &'a mut usize,
+            ) -> impl FnMut((), T) -> ControlFlow<usize, ()> + 'a {
+                move |_, x| {
+                    if predicate(x) {
+                        ControlFlow::Break(*acc)
+                    } else {
+                        *acc += 1;
+                        ControlFlow::Continue(())
+                    }
+                }
+            }
+
+            let mut acc = 0;
+            match iter.try_fold((), check(predicate, &mut acc)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
 }
 
 impl ExactSizeIterator for BoundDictIterator<'_> {
     fn len(&self) -> usize {
-        self.len as usize
+        match self {
+            BoundDictIterator::ItemIter { remaining, .. } => *remaining as usize,
+            BoundDictIterator::DictIter { remaining, .. } => *remaining as usize,
+        }
     }
 }
 
 impl<'py> BoundDictIterator<'py> {
     fn new(dict: Bound<'py, PyDict>) -> Self {
-        let len = dict_len(&dict);
-        BoundDictIterator {
-            dict,
-            ppos: 0,
-            di_used: len,
-            len,
+        let remaining = dict_len(&dict);
+
+        if dict.is_exact_instance_of::<PyDict>() {
+            return BoundDictIterator::DictIter {
+                dict,
+                ppos: 0,
+                di_used: remaining,
+                remaining,
+            };
+        };
+
+        let items = dict.call_method0(intern!(dict.py(), "items")).unwrap();
+        let iter = PyIterator::from_object(&items).unwrap();
+        BoundDictIterator::ItemIter { iter, remaining }
+    }
+
+    #[inline]
+    fn with_critical_section<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut Self) -> R,
+    {
+        match self {
+            BoundDictIterator::ItemIter { .. } => f(self),
+            #[cfg(not(Py_GIL_DISABLED))]
+            BoundDictIterator::DictIter { .. } => f(self),
+            #[cfg(Py_GIL_DISABLED)]
+            BoundDictIterator::DictIter { ref dict, .. } => {
+                crate::sync::with_critical_section(dict.clone().as_ref(), || f(self))
+            }
```

Comment:
Think this can be simplified given that `with_critical_section` should optimize away trivially on the gil-enabled build, also can remove a clone:

```suggestion
            BoundDictIterator::DictIter { dict } => {
                crate::sync::with_critical_section(dict.as_any(), || f(self))
            }
```

Reply:
I don't think the borrow checker will allow this because we need a mut reference to `self` while iterating. So it will not let is have a second reference without Copy to get a strong ref to Bound.

Reply:
Ah, good point. I think if we split the `BoundDictIterator` into the main struct and a private `inner`, we might be able to make it work? That's potentially a good idea anyway so that we don't have to make the enum members public, and it would be desirable to avoid the redundant `.clone()` as a performance optimization :)

Reply:
(So probably gets simpler with https://github.com/PyO3/pyo3/pull/4439#discussion_r1807471142)